api/envoy/extensions/transport_sockets/tls/v3/tls.proto

```diff
@@ -82,6 +82,19 @@ message UpstreamTlsContext {
   // The ``ssl.was_key_usage_invalid`` in :ref:`listener metrics <config_listener_stats>` metric will be incremented
   // for configurations that would fail if this option were enabled.
   google.protobuf.BoolValue enforce_rsa_key_usage = 5;
+
+  // Enables caching of TLS session keys on a per-upstream host basis.
+  // This field specifies the maximum number of upstream hosts for which TLS session keys
+  // will be cached. It is particularly useful when upstream hosts serving the same destination
+  // have unique TLS configurations, meaning that session keys for one host cannot be used with another.
+  //
+  // Note: The `max_session_keys` field (which limits the number of session keys per upstream)
+  // works in conjunction with this setting. For instance, if `max_session_keys` is 10 and
+  // `max_session_cache_upstream_hosts` is 20, the total number of session keys in the cache
+  // would be 200.
+  //
+  // Defaults to 0, which disables TLS session key caching per upstream host.
+  google.protobuf.UInt32Value max_session_cache_upstream_hosts = 8;
```

Comment:
Is it possible that there will be other configurable parameters for per-upstream-host session caching in the future?  If so, consider using a separate message as an alternative:

```
message TlsPerHostSessionCacheConfig {
  uint32 max_session_cache_upstream_hosts = 1;

  // ...can add more fields here in the future...
}

message UpstreamTlsContext {
  // ...existing fields...

  // If set, per-host session caching is enabled.
  TlsPerHostSessionCacheConfig per_host_session_cache_config = 8;
}
```

Reply:
Yeah it could potentially carry other features in the future, updated with separate message.